File: homework1/17_Letter_Combinations_of_a_Phone_Number.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
using namespace std;


class Solution {
public:
    vector<string> phone = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> result;
    string path;

    void func(const string& digital, int index){
        // dk dừng
        if ( index == digital.size()){
            result.push_back(path);
            return;
        }

        // ghép từ
        for ( char ch : phone[digital[index] - '0']){
            path += ch;
            func(digital, index +1);
            path.pop_back();
        }
    }


    vector<string> letterCombinations(string digits) {
        func(digits, 0);
        return result;
    }
};

int main(){
    Solution sol;
    string digits = "23";
    for ( string x : sol.letterCombinations(digits)){
        cout << x << endl;
    }
}

File: homework1/17_Letter_Combinations_of_a_Phone_Number.py
from typing import List
class Solution:
    def __init__(self):
        self.phone = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        self.result = []
        self.path = ""

    def func(self, digits:str, index):
        # dk dung
        if index == len(digits):
            self.result.append(self.path);
            return

        for ch in self.phone[int(digits[index])]:
            self.path += ch
            self.func(digits, index +1)
            self.path = self.path[:-1]


    def letterCombinations(self, digits: str) -> List[str]:
        self.func(digits, 0)
        return self.result

